# EBAC - Regressão II - regressão múltipla

## Tarefa I

#### Previsão de renda

Vamos trabalhar com a base 'previsao_de_renda.csv', que é a base do seu próximo projeto. Vamos usar os recursos que vimos até aqui nesta base.

|variavel|descrição|
|-|-|
|data_ref                | Data de referência de coleta das variáveis |
|index                   | Código de identificação do cliente|
|sexo                    | Sexo do cliente|
|posse_de_veiculo        | Indica se o cliente possui veículo|
|posse_de_imovel         | Indica se o cliente possui imóvel|
|qtd_filhos              | Quantidade de filhos do cliente|
|tipo_renda              | Tipo de renda do cliente|
|educacao                | Grau de instrução do cliente|
|estado_civil            | Estado civil do cliente|
|tipo_residencia         | Tipo de residência do cliente (própria, alugada etc)|
|idade                   | Idade do cliente|
|tempo_emprego           | Tempo no emprego atual|
|qt_pessoas_residencia   | Quantidade de pessoas que moram na residência|
|renda                   | Renda em reais|

In [34]:
import pandas as pd
import seaborn as sns
from seaborn import load_dataset
import matplotlib.pyplot as plt
import numpy as np

import patsy
import statsmodels.api as sm
import statsmodels.formula.api as smf

%matplotlib inline

In [35]:
df = pd.read_csv('/content/previsao_de_renda.csv')

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             15000 non-null  int64  
 1   data_ref               15000 non-null  object 
 2   id_cliente             15000 non-null  int64  
 3   sexo                   15000 non-null  object 
 4   posse_de_veiculo       15000 non-null  bool   
 5   posse_de_imovel        15000 non-null  bool   
 6   qtd_filhos             15000 non-null  int64  
 7   tipo_renda             15000 non-null  object 
 8   educacao               15000 non-null  object 
 9   estado_civil           15000 non-null  object 
 10  tipo_residencia        15000 non-null  object 
 11  idade                  15000 non-null  int64  
 12  tempo_emprego          12427 non-null  float64
 13  qt_pessoas_residencia  15000 non-null  float64
 14  renda                  15000 non-null  float64
dtypes:

1. Ajuste um modelo para prever log(renda) considerando todas as covariáveis disponíveis.
    - Utilizando os recursos do Patsy, coloque as variáveis qualitativas como *dummies*.
    - Mantenha sempre a categoria mais frequente como casela de referência
    - Avalie os parâmetros e veja se parecem fazer sentido prático.  



In [37]:
df.head()

Unnamed: 0    data_ref  id_cliente sexo  posse_de_veiculo  posse_de_imovel  \
0           0  2015-01-01       15056    F             False             True   
1           1  2015-01-01        9968    M              True             True   
2           2  2015-01-01        4312    F              True             True   
3           3  2015-01-01       10639    F             False             True   
4           4  2015-01-01        7064    M              True            False   

   qtd_filhos        tipo_renda           educacao estado_civil  \
0           0        Empresário         Secundário     Solteiro   
1           0       Assalariado  Superior completo       Casado   
2           0        Empresário  Superior completo       Casado   
3           1  Servidor público  Superior completo       Casado   
4           0       Assalariado         Secundário     Solteiro   

  tipo_residencia  idade  tempo_emprego  qt_pessoas_residencia    renda  
0            Casa     26       6.602740                    1.0  8060.34  
1            Casa     28       7.183562                    2.0  1852.15  
2            Casa     35       0.838356                    2.0  2253.89  
3            Casa     30       4.846575                    3.0  6600.77  
4   Governamental     33       4.293151                    1.0  6475.97

In [66]:
y, x =  patsy.dmatrices('np.log(renda) ~ C(sexo) + C(posse_de_veiculo) + C(posse_de_imovel) + qtd_filhos + C(tipo_renda) + C(educacao) + C(estado_civil) + C(tipo_residencia) + idade + tempo_emprego + qt_pessoas_residencia', data = df)
x

DesignMatrix with shape (12427, 25)
  Columns:
    ['Intercept',
     'C(sexo)[T.M]',
     'C(posse_de_veiculo)[T.True]',
     'C(posse_de_imovel)[T.True]',
     'C(tipo_renda)[T.Bolsista]',
     'C(tipo_renda)[T.Empresário]',
     'C(tipo_renda)[T.Pensionista]',
     'C(tipo_renda)[T.Servidor público]',
     'C(educacao)[T.Pós graduação]',
     'C(educacao)[T.Secundário]',
     'C(educacao)[T.Superior completo]',
     'C(educacao)[T.Superior incompleto]',
     'C(estado_civil)[T.Separado]',
     'C(estado_civil)[T.Solteiro]',
     'C(estado_civil)[T.União]',
     'C(estado_civil)[T.Viúvo]',
     'C(tipo_residencia)[T.Casa]',
     'C(tipo_residencia)[T.Com os pais]',
     'C(tipo_residencia)[T.Comunitário]',
     'C(tipo_residencia)[T.Estúdio]',
     'C(tipo_residencia)[T.Governamental]',
     'qtd_filhos',
     'idade',
     'tempo_emprego',
     'qt_pessoas_residencia']
  Terms:
    'Intercept' (column 0)
    'C(sexo)' (column 1)
    'C(posse_de_veiculo)' (column 2)
    'C(posse_de_i

In [70]:
print(sm.OLS(y, x).fit().summary())

                            OLS Regression Results                            
Dep. Variable:          np.log(renda)   R-squared:                       0.357
Model:                            OLS   Adj. R-squared:                  0.356
Method:                 Least Squares   F-statistic:                     287.5
Date:                Fri, 17 Feb 2023   Prob (F-statistic):               0.00
Time:                        18:16:35   Log-Likelihood:                -13568.
No. Observations:               12427   AIC:                         2.719e+04
Df Residuals:                   12402   BIC:                         2.737e+04
Df Model:                          24                                         
Covariance Type:            nonrobust                                         
                                          coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------
In

In [74]:
df[['sexo']].value_counts()

sexo
F       10119
M        4881
dtype: int64

In [77]:
df[['posse_de_veiculo']].value_counts()

posse_de_veiculo
False               9140
True                5860
dtype: int64

In [79]:
df[['posse_de_imovel']].value_counts()

posse_de_imovel
True               10143
False               4857
dtype: int64

In [80]:
df[['tipo_renda']].value_counts()

tipo_renda      
Assalariado         7633
Empresário          3508
Pensionista         2582
Servidor público    1268
Bolsista               9
dtype: int64

In [83]:
df[['educacao']].value_counts()

educacao           
Secundário             8895
Superior completo      5335
Superior incompleto     579
Primário                165
Pós graduação            26
dtype: int64

In [87]:
df[['estado_civil']].value_counts()

estado_civil
Casado          10534
Solteiro         1798
União            1078
Separado          879
Viúvo             711
dtype: int64

In [89]:
df[['tipo_residencia']].value_counts()

tipo_residencia
Casa               13532
Com os pais          675
Governamental        452
Aluguel              194
Estúdio               83
Comunitário           64
dtype: int64

In [92]:
y, x =  patsy.dmatrices('np.log(renda) ~ C(sexo, Treatment(1)) + C(posse_de_veiculo, Treatment(1)) + C(posse_de_imovel) + qtd_filhos + C(tipo_renda, Treatment(0)) + C(educacao, Treatment(2)) + C(estado_civil, Treatment(0)) + C(tipo_residencia, Treatment(1)) + idade + tempo_emprego + qt_pessoas_residencia', data = df)
x

DesignMatrix with shape (12427, 25)
  Columns:
    ['Intercept',
     'C(sexo, Treatment(1))[T.F]',
     'C(posse_de_veiculo, Treatment(1))[T.False]',
     'C(posse_de_imovel)[T.True]',
     'C(tipo_renda, Treatment(0))[T.Bolsista]',
     'C(tipo_renda, Treatment(0))[T.Empresário]',
     'C(tipo_renda, Treatment(0))[T.Pensionista]',
     'C(tipo_renda, Treatment(0))[T.Servidor público]',
     'C(educacao, Treatment(2))[T.Primário]',
     'C(educacao, Treatment(2))[T.Pós graduação]',
     'C(educacao, Treatment(2))[T.Superior completo]',
     'C(educacao, Treatment(2))[T.Superior incompleto]',
     'C(estado_civil, Treatment(0))[T.Separado]',
     'C(estado_civil, Treatment(0))[T.Solteiro]',
     'C(estado_civil, Treatment(0))[T.União]',
     'C(estado_civil, Treatment(0))[T.Viúvo]',
     'C(tipo_residencia, Treatment(1))[T.Aluguel]',
     'C(tipo_residencia, Treatment(1))[T.Com os pais]',
     'C(tipo_residencia, Treatment(1))[T.Comunitário]',
     'C(tipo_residencia, Treatment(1))[T.E

In [93]:
print(sm.OLS(y, x).fit().summary())

                            OLS Regression Results                            
Dep. Variable:          np.log(renda)   R-squared:                       0.357
Model:                            OLS   Adj. R-squared:                  0.356
Method:                 Least Squares   F-statistic:                     287.5
Date:                Fri, 17 Feb 2023   Prob (F-statistic):               0.00
Time:                        18:53:04   Log-Likelihood:                -13568.
No. Observations:               12427   AIC:                         2.719e+04
Df Residuals:                   12402   BIC:                         2.737e+04
Df Model:                          24                                         
Covariance Type:            nonrobust                                         
                                                        coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------

*Ajustando o modelo com as caselas mais frequentes ainda não vemos uma grande mudança nos P>|t|. Talvez quando retirarmos algumas variáveis possa melhorar o modelo.*


2. Remova a variável menos significante e analise:
    - Observe os indicadores que vimos, e avalie se o modelo melhorou ou piorou na sua opinião.
    - Observe os parâmetros e veja se algum se alterou muito.  



    

In [110]:
y, x =  patsy.dmatrices('np.log(renda) ~ C(sexo, Treatment(1)) + C(posse_de_veiculo, Treatment(1)) + C(posse_de_imovel) + qtd_filhos + C(tipo_renda, Treatment(0)) + idade + tempo_emprego +  C(educacao, Treatment(2))', data = df)
x

DesignMatrix with shape (12427, 15)
  Columns:
    ['Intercept',
     'C(sexo, Treatment(1))[T.F]',
     'C(posse_de_veiculo, Treatment(1))[T.False]',
     'C(posse_de_imovel)[T.True]',
     'C(tipo_renda, Treatment(0))[T.Bolsista]',
     'C(tipo_renda, Treatment(0))[T.Empresário]',
     'C(tipo_renda, Treatment(0))[T.Pensionista]',
     'C(tipo_renda, Treatment(0))[T.Servidor público]',
     'C(educacao, Treatment(2))[T.Primário]',
     'C(educacao, Treatment(2))[T.Pós graduação]',
     'C(educacao, Treatment(2))[T.Superior completo]',
     'C(educacao, Treatment(2))[T.Superior incompleto]',
     'qtd_filhos',
     'idade',
     'tempo_emprego']
  Terms:
    'Intercept' (column 0)
    'C(sexo, Treatment(1))' (column 1)
    'C(posse_de_veiculo, Treatment(1))' (column 2)
    'C(posse_de_imovel)' (column 3)
    'C(tipo_renda, Treatment(0))' (columns 4:8)
    'C(educacao, Treatment(2))' (columns 8:12)
    'qtd_filhos' (column 12)
    'idade' (column 13)
    'tempo_emprego' (column 14)
  (

In [111]:
print(sm.OLS(y, x).fit().summary())

                            OLS Regression Results                            
Dep. Variable:          np.log(renda)   R-squared:                       0.356
Model:                            OLS   Adj. R-squared:                  0.356
Method:                 Least Squares   F-statistic:                     491.1
Date:                Fri, 17 Feb 2023   Prob (F-statistic):               0.00
Time:                        19:43:38   Log-Likelihood:                -13577.
No. Observations:               12427   AIC:                         2.718e+04
Df Residuals:                   12412   BIC:                         2.730e+04
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
                                                       coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------

*A grande mudança foi no F-statistic do modelo inicial, para o modelo com variáveis reduzidas. 287.5 versus 491.1. *

3. Siga removendo as variáveis menos significantes, sempre que o *p-value* for menor que 5%. Compare o modelo final com o inicial. Observe os indicadores e conclua se o modelo parece melhor. 

In [122]:
y, x =  patsy.dmatrices('np.log(renda) ~ C(sexo, Treatment(1)) + C(posse_de_veiculo, Treatment(1)) + C(posse_de_imovel) + C(tipo_renda) + qtd_filhos + idade + tempo_emprego', data = df)
x

DesignMatrix with shape (12427, 11)
  Columns:
    ['Intercept',
     'C(sexo, Treatment(1))[T.F]',
     'C(posse_de_veiculo, Treatment(1))[T.False]',
     'C(posse_de_imovel)[T.True]',
     'C(tipo_renda)[T.Bolsista]',
     'C(tipo_renda)[T.Empresário]',
     'C(tipo_renda)[T.Pensionista]',
     'C(tipo_renda)[T.Servidor público]',
     'qtd_filhos',
     'idade',
     'tempo_emprego']
  Terms:
    'Intercept' (column 0)
    'C(sexo, Treatment(1))' (column 1)
    'C(posse_de_veiculo, Treatment(1))' (column 2)
    'C(posse_de_imovel)' (column 3)
    'C(tipo_renda)' (columns 4:8)
    'qtd_filhos' (column 8)
    'idade' (column 9)
    'tempo_emprego' (column 10)
  (to view full data, use np.asarray(this_obj))

In [123]:
print(sm.OLS(y, x).fit().summary())

                            OLS Regression Results                            
Dep. Variable:          np.log(renda)   R-squared:                       0.353
Model:                            OLS   Adj. R-squared:                  0.353
Method:                 Least Squares   F-statistic:                     677.5
Date:                Fri, 17 Feb 2023   Prob (F-statistic):               0.00
Time:                        20:43:05   Log-Likelihood:                -13610.
No. Observations:               12427   AIC:                         2.724e+04
Df Residuals:                   12416   BIC:                         2.732e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                                                 coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------

*O F-statistic melhora muito. Acabei não tirando a variável 'tipo_renda' pois em duas vertenter ela tem um p-value menor que 5%.*